# SQL Exercise
Crea notebook de Jupyter y con la librería de SQLite realiza las siguientes consultas a la base de datos "Chinook_Sqlite" (archivo proporcionado junto con los apuntes de SQL). columnas.

In [1]:
import sqlite3
import pandas as pd

In [2]:
# Connect to the database
con = sqlite3.connect("data/Chinook_Sqlite.sqlite") 

# Create a cursor
cursor = con.cursor()

## Actividad 2-1 (Atención al cliente)

Trabaja con la tabla Customer.

- Muestra la columna Email.
- Muestra la columna FirstName.
- Muestra las columnas FirstName y LastName.
- Muestra las columnas Address, Phone y Fax.
- Muestra todas las columnas.

**1. Muestra la columna Email.**

In [3]:
query = "SELECT Email FROM Customer"

# Run the query and get the results
cursor.execute(query)
emails = cursor.fetchall()

# Convert the results to a DataFrame of pandas for better visualization.
emails_df = pd.DataFrame(emails, columns=['Email'])

emails_df.head(10)

Email
0      luisg@embraer.com.br
1     leonekohler@surfeu.de
2       ftremblay@gmail.com
3     bjorn.hansen@yahoo.no
4  frantisekw@jetbrains.com
5           hholy@gmail.com
6    astrid.gruber@apple.at
7     daan_peeters@apple.be
8     kara.nielsen@jubii.dk
9  eduardo@woodstock.com.br

**2. Muestra la columna FirstName.**

In [4]:
query = "SELECT FirstName FROM Customer"

cursor.execute(query)
first_name = cursor.fetchall()

first_name_df = pd.DataFrame(first_name, columns=["First Name"])
first_name_df.head(10)

First Name
0       Luís
1     Leonie
2   François
3      Bjørn
4  František
5     Helena
6     Astrid
7       Daan
8       Kara
9    Eduardo

**3. Muestra las columnas FirstName y LastName.**

In [5]:
query = "SELECT FirstName, LastName FROM Customer"

cursor.execute(query)
full_name = cursor.fetchall()

full_name_df = pd.DataFrame(full_name, columns=["First Name", "Last Name"])
full_name_df.head()

First Name    Last Name
0       Luís    Gonçalves
1     Leonie       Köhler
2   François     Tremblay
3      Bjørn       Hansen
4  František  Wichterlová

**4. Muestra las columnas Address, Phone y Fax.**

In [6]:
query = "SELECT Address, Phone, Fax FROM Customer"

cursor.execute(query)
adress = cursor.fetchall()

address_df = pd.DataFrame(adress, columns=["Address", "Phone", "Fax"])
address_df.head(10)

Address               Phone  \
0       Av. Brigadeiro Faria Lima, 2170  +55 (12) 3923-5555   
1               Theodor-Heuss-Straße 34    +49 0711 2842222   
2                     1498 rue Bélanger   +1 (514) 721-4711   
3                      Ullevålsveien 14     +47 22 44 22 22   
4                         Klanova 9/506    +420 2 4172 5555   
5                         Rilská 3174/6    +420 2 4177 0449   
6  Rotenturmstraße 4, 1010 Innere Stadt      +43 01 5134505   
7                       Grétrystraat 63    +32 02 219 03 03   
8                   Sønder Boulevard 51      +453 3331 9991   
9             Rua Dr. Falcão Filho, 155  +55 (11) 3033-5446   

                  Fax  
0  +55 (12) 3923-5566  
1                None  
2                None  
3                None  
4    +420 2 4172 5555  
5                None  
6                None  
7                None  
8                None  
9  +55 (11) 3033-4564

**5. Muestra todas las columnas.**

In [7]:
query = "SELECT * FROM Customer"

cursor.execute(query)
customer = cursor.fetchall()

column_names = [description[0] for description in cursor.description]

customer_df = pd.DataFrame(customer, columns=[column_names])
customer_df.head()

CustomerId  FirstName     LastName  \
0          1       Luís    Gonçalves   
1          2     Leonie       Köhler   
2          3   François     Tremblay   
3          4      Bjørn       Hansen   
4          5  František  Wichterlová   

                                            Company  \
0  Embraer - Empresa Brasileira de Aeronáutica S.A.   
1                                              None   
2                                              None   
3                                              None   
4                                  JetBrains s.r.o.   

                           Address                 City State         Country  \
0  Av. Brigadeiro Faria Lima, 2170  São José dos Campos    SP          Brazil   
1          Theodor-Heuss-Straße 34            Stuttgart  None         Germany   
2                1498 rue Bélanger             Montréal    QC          Canada   
3                 Ullevålsveien 14                 Oslo  None          Norway   
4                    Klanova 9/506               Prague  None  Czech Republic   

  PostalCode               Phone                 Fax  \
0  12227-000  +55 (12) 3923-5555  +55 (12) 3923-5566   
1      70174    +49 0711 2842222                None   
2    H2G 1A7   +1 (514) 721-4711                None   
3       0171     +47 22 44 22 22                None   
4      14700    +420 2 4172 5555    +420 2 4172 5555   

                      Email SupportRepId  
0      luisg@embraer.com.br            3  
1     leonekohler@surfeu.de            5  
2       ftremblay@gmail.com            3  
3     bjorn.hansen@yahoo.no            4  
4  frantisekw@jetbrains.com            4

## Actividad 3-1 (Ordenando a mi manera)
Ordena la tabla de géneros (Genre) por orden alfabético.

Corta los resultados para que solo se muestren los 6 primeros.

Subir Nota:
- Muestra las últimas 2 filas de la tabla.

**1. Ordena la tabla de géneros (Genre) por orden alfabético.**

In [8]:
query = "SELECT * FROM Genre ORDER BY Name ASC"

cursor.execute(query)
genre = cursor.fetchall()

genre_df = pd.DataFrame(genre, columns=["GenreId", "Name"])
genre_df.head(10)

GenreId                Name
0       23         Alternative
1        4  Alternative & Punk
2        6               Blues
3       11          Bossa Nova
4       24           Classical
5       22              Comedy
6       21               Drama
7       12      Easy Listening
8       15   Electronica/Dance
9       13         Heavy Metal

**2. Corta los resultados para que solo se muestren los 6 primeros.**

In [9]:
query = "SELECT * FROM Genre ORDER BY Name ASC LIMIT 6 OFFSET 0"

cursor.execute(query)
genre = cursor.fetchall()

genre_df = pd.DataFrame(genre, columns=["GenreId", "Name"])
genre_df

GenreId                Name
0       23         Alternative
1        4  Alternative & Punk
2        6               Blues
3       11          Bossa Nova
4       24           Classical
5       22              Comedy

**3. Muestra las últimas 2 filas de la tabla.**

In [10]:
# SOLUTION WITH PYTHON
query = "SELECT * FROM Genre ORDER BY Name ASC"

cursor.execute(query)
genre = cursor.fetchall()

genre_df = pd.DataFrame(genre, columns=["GenreId", "Name"])
print("Últimas 2 filas de la tabla Genre ordenada alfabéticamente:")
genre_df.tail(2)

Últimas 2 filas de la tabla Genre ordenada alfabéticamente:


GenreId      Name
23       19  TV Shows
24       16     World

In [11]:
# SQL SOLUTION
query = """
SELECT * FROM (
    SELECT * FROM Genre ORDER BY Name ASC
) ORDER BY Name DESC LIMIT 2;
"""
cursor.execute(query)
genre = cursor.fetchall()

genre_df = pd.DataFrame(genre, columns=["GenreId", "Name"])
genre_df

GenreId      Name
0       16     World
1       19  TV Shows

## Actividad 3-2 (En busca del peor cliente)
Volvemos a usar la tabla Customer.

* Muestra los usuarios que han reportado más de 3 incidencia (columna donde se cuentan es SupportRepid)
* Filtra los resultados anteriores por los que viven en Brasil.
* Muestra a los usuarios que tengan un código postal empezado en 7.
* Muestra a los usuarios que tengan un email de hotmail.
* Muestra los usuarios nacidos en Estados Unidos (USA) o Canadá (Canada).
* De los resultados anteriores, muestra los que tengan un email de gmail.
* Muestra al usuario que trabaja en Apple (columna Company). Te aviso que no sabes el formato de la compañía, podría ser: Apple SL, Company Apple, APPLE…
* Muestra los usuario que han reportado entre 3 y 4 incidencias.

**1. Muestra los usuarios que han reportado más de 3 incidencia (columna donde se cuentan es SupportRepid)**

In [12]:
query = "SELECT * FROM Customer WHERE SupportRepid == 3"

cursor.execute(query)
incidence = cursor.fetchall()

column_names = [description[0] for description in cursor.description]
incidence_df = pd.DataFrame(incidence, columns=column_names)
incidence_df.head()

CustomerId FirstName   LastName  \
0           1      Luís  Gonçalves   
1           3  François   Tremblay   
2          12   Roberto    Almeida   
3          15  Jennifer   Peterson   
4          18  Michelle     Brooks   

                                            Company  \
0  Embraer - Empresa Brasileira de Aeronáutica S.A.   
1                                              None   
2                                            Riotur   
3                                     Rogers Canada   
4                                              None   

                           Address                 City State Country  \
0  Av. Brigadeiro Faria Lima, 2170  São José dos Campos    SP  Brazil   
1                1498 rue Bélanger             Montréal    QC  Canada   
2                 Praça Pio X, 119       Rio de Janeiro    RJ  Brazil   
3              700 W Pender Street            Vancouver    BC  Canada   
4                     627 Broadway             New York    NY     USA   

   PostalCode               Phone                 Fax  \
0   12227-000  +55 (12) 3923-5555  +55 (12) 3923-5566   
1     H2G 1A7   +1 (514) 721-4711                None   
2   20040-020  +55 (21) 2271-7000  +55 (21) 2271-7070   
3     V6C 1G8   +1 (604) 688-2255   +1 (604) 688-8756   
4  10012-2612   +1 (212) 221-3546   +1 (212) 221-4679   

                           Email  SupportRepId  
0           luisg@embraer.com.br             3  
1            ftremblay@gmail.com             3  
2  roberto.almeida@riotur.gov.br             3  
3            jenniferp@rogers.ca             3  
4              michelleb@aol.com             3

**2. Filtra los resultados anteriores por los que viven en Brasil.**

In [13]:
query = "SELECT * FROM Customer WHERE SupportRepId = 3 AND Country = 'Brazil'"

cursor.execute(query)
incidence_brazil = cursor.fetchall()

incidence_df = pd.DataFrame(incidence_brazil, columns=column_names)
incidence_df

CustomerId FirstName   LastName  \
0           1      Luís  Gonçalves   
1          12   Roberto    Almeida   

                                            Company  \
0  Embraer - Empresa Brasileira de Aeronáutica S.A.   
1                                            Riotur   

                           Address                 City State Country  \
0  Av. Brigadeiro Faria Lima, 2170  São José dos Campos    SP  Brazil   
1                 Praça Pio X, 119       Rio de Janeiro    RJ  Brazil   

  PostalCode               Phone                 Fax  \
0  12227-000  +55 (12) 3923-5555  +55 (12) 3923-5566   
1  20040-020  +55 (21) 2271-7000  +55 (21) 2271-7070   

                           Email  SupportRepId  
0           luisg@embraer.com.br             3  
1  roberto.almeida@riotur.gov.br             3

**3. Muestra a los usuarios que tengan un código postal empezado en 7.**

In [14]:
query = "SELECT * FROM Customer WHERE PostalCode LIKE '7%'"

cursor.execute(query)
postal_code = cursor.fetchall()

postal_code = pd.DataFrame(postal_code, columns=column_names)
postal_code

CustomerId  FirstName    LastName Company                  Address  \
0           2     Leonie      Köhler    None  Theodor-Heuss-Straße 34   
1          13   Fernanda       Ramos    None             Qe 7 Bloco G   
2          26    Richard  Cunningham    None      2211 W Berry Street   
3          39    Camille     Bernard    None            4, Rue Milton   
4          40  Dominique    Lefebvre    None           8, Rue Hanovre   

         City State  Country PostalCode               Phone  \
0   Stuttgart  None  Germany      70174    +49 0711 2842222   
1    Brasília    DF   Brazil  71020-677  +55 (61) 3363-5547   
2  Fort Worth    TX      USA      76110   +1 (817) 924-7272   
3       Paris  None   France      75009  +33 01 49 70 65 65   
4       Paris  None   France      75002  +33 01 47 42 71 71   

                  Fax                        Email  SupportRepId  
0                None        leonekohler@surfeu.de             5  
1  +55 (61) 3363-7855     fernadaramos4@uol.com.br             4  
2                None     ricunningham@hotmail.com             4  
3                None     camille.bernard@yahoo.fr             4  
4                None  dominiquelefebvre@gmail.com             4

**4. Muestra a los usuarios que tengan un email de hotmail.**

In [15]:
query = "SELECT * FROM Customer WHERE Email LIKE '%@hotmail.com'"

cursor.execute(query)
email = cursor.fetchall()

email = pd.DataFrame(email, columns=column_names)
email

CustomerId FirstName    LastName Company              Address        City  \
0          21     Kathy       Chase    None     801 W 4th Street        Reno   
1          26   Richard  Cunningham    None  2211 W Berry Street  Fort Worth   
2          41      Marc      Dubois    None  11, Place Bellecour        Lyon   
3          52      Emma       Jones    None    202 Hoxton Street      London   

  State         Country PostalCode               Phone   Fax  \
0    NV             USA      89503   +1 (775) 223-7665  None   
1    TX             USA      76110   +1 (817) 924-7272  None   
2  None          France      69002  +33 04 78 30 30 30  None   
3  None  United Kingdom     N1 5LH   +44 020 7707 0707  None   

                      Email  SupportRepId  
0       kachase@hotmail.com             5  
1  ricunningham@hotmail.com             4  
2   marc.dubois@hotmail.com             5  
3    emma_jones@hotmail.com             3

**5. Muestra los usuarios nacidos en Estados Unidos (USA) o Canadá (Canada).**

In [16]:
query = "SELECT * FROM Customer WHERE Country == 'USA'  OR Country == 'Canada'"

cursor.execute(query)
usa_cnd = cursor.fetchall()

usa_cnd_df = pd.DataFrame(usa_cnd, columns=column_names)
usa_cnd_df

CustomerId FirstName    LastName                Company  \
0            3  François    Tremblay                   None   
1           14      Mark     Philips                  Telus   
2           15  Jennifer    Peterson          Rogers Canada   
3           16     Frank      Harris            Google Inc.   
4           17      Jack       Smith  Microsoft Corporation   
5           18  Michelle      Brooks                   None   
6           19       Tim       Goyer             Apple Inc.   
7           20       Dan      Miller                   None   
8           21     Kathy       Chase                   None   
9           22   Heather     Leacock                   None   
10          23      John      Gordon                   None   
11          24     Frank     Ralston                   None   
12          25    Victor     Stevens                   None   
13          26   Richard  Cunningham                   None   
14          27   Patrick        Gray                   None   
15          28     Julia     Barnett                   None   
16          29    Robert       Brown                   None   
17          30    Edward     Francis                   None   
18          31    Martha        Silk                   None   
19          32     Aaron    Mitchell                   None   
20          33     Ellie    Sullivan                   None   

                      Address            City State Country  PostalCode  \
0           1498 rue Bélanger        Montréal    QC  Canada     H2G 1A7   
1              8210 111 ST NW        Edmonton    AB  Canada     T6G 2C7   
2         700 W Pender Street       Vancouver    BC  Canada     V6C 1G8   
3   1600 Amphitheatre Parkway   Mountain View    CA     USA  94043-1351   
4             1 Microsoft Way         Redmond    WA     USA  98052-8300   
5                627 Broadway        New York    NY     USA  10012-2612   
6             1 Infinite Loop       Cupertino    CA     USA       95014   
7        541 Del Medio Avenue   Mountain View    CA     USA   94040-111   
8            801 W 4th Street            Reno    NV     USA       89503   
9            120 S Orange Ave         Orlando    FL     USA       32801   
10            69 Salem Street          Boston    MA     USA        2113   
11      162 E Superior Street         Chicago    IL     USA       60611   
12      319 N. Frances Street         Madison    WI     USA       53703   
13        2211 W Berry Street      Fort Worth    TX     USA       76110   
14            1033 N Park Ave          Tucson    AZ     USA       85719   
15                302 S 700 E  Salt Lake City    UT     USA       84102   
16     796 Dundas Street West         Toronto    ON  Canada     M6J 1V1   
17           230 Elgin Street          Ottawa    ON  Canada     K2P 1L7   
18      194A Chain Lake Drive         Halifax    NS  Canada     B3S 1C5   
19         696 Osborne Street        Winnipeg    MB  Canada     R3L 2B9   
20             5112 48 Street     Yellowknife    NT  Canada     X1A 1N6   

                Phone                Fax                     Email  \
0   +1 (514) 721-4711               None       ftremblay@gmail.com   
1   +1 (780) 434-4554  +1 (780) 434-5565        mphilips12@shaw.ca   
2   +1 (604) 688-2255  +1 (604) 688-8756       jenniferp@rogers.ca   
3   +1 (650) 253-0000  +1 (650) 253-0000        fharris@google.com   
4   +1 (425) 882-8080  +1 (425) 882-8081   jacksmith@microsoft.com   
5   +1 (212) 221-3546  +1 (212) 221-4679         michelleb@aol.com   
6   +1 (408) 996-1010  +1 (408) 996-1011          tgoyer@apple.com   
7   +1 (650) 644-3358               None       dmiller@comcast.com   
8   +1 (775) 223-7665               None       kachase@hotmail.com   
9   +1 (407) 999-7788               None        hleacock@gmail.com   
10  +1 (617) 522-1333               None    johngordon22@yahoo.com   
11  +1 (312) 332-3232               None        fralston@gmail.com   
12  +1 (608) 257-0597               None        vstevens

**6. De los resultados anteriores, muestra los que tengan un email de gmail.**

In [17]:
query = """
SELECT * FROM Customer 
WHERE (Country == 'USA' OR Country == 'Canada') AND Email LIKE '%@gmail.%'
"""
cursor.execute(query)
usa_cnd = cursor.fetchall()

usa_cnd_df = pd.DataFrame(usa_cnd, columns=column_names)
usa_cnd_df

CustomerId FirstName  LastName Company                Address  \
0           3  François  Tremblay    None      1498 rue Bélanger   
1          22   Heather   Leacock    None       120 S Orange Ave   
2          24     Frank   Ralston    None  162 E Superior Street   
3          28     Julia   Barnett    None            302 S 700 E   
4          31    Martha      Silk    None  194A Chain Lake Drive   

             City State Country PostalCode              Phone   Fax  \
0        Montréal    QC  Canada    H2G 1A7  +1 (514) 721-4711  None   
1         Orlando    FL     USA      32801  +1 (407) 999-7788  None   
2         Chicago    IL     USA      60611  +1 (312) 332-3232  None   
3  Salt Lake City    UT     USA      84102  +1 (801) 531-7272  None   
4         Halifax    NS  Canada    B3S 1C5  +1 (902) 450-0450  None   

                  Email  SupportRepId  
0   ftremblay@gmail.com             3  
1    hleacock@gmail.com             4  
2    fralston@gmail.com             3  
3   jubarnett@gmail.com             5  
4  marthasilk@gmail.com             5

**7. Muestra al usuario que trabaja en Apple (columna Company). Te aviso que no sabes el formato de la compañía, podría ser: Apple SL, Company Apple, APPLE…**

In [18]:
query = "SELECT * FROM Customer WHERE Company LIKE '%apple%' COLLATE NOCASE"

cursor.execute(query)
company = cursor.fetchall()

company_df = pd.DataFrame(company, columns=column_names)
company_df

CustomerId FirstName LastName     Company          Address       City  \
0          19       Tim    Goyer  Apple Inc.  1 Infinite Loop  Cupertino   

  State Country PostalCode              Phone                Fax  \
0    CA     USA      95014  +1 (408) 996-1010  +1 (408) 996-1011   

              Email  SupportRepId  
0  tgoyer@apple.com             3

**8. Muestra los usuario que han reportado entre 3 y 4 incidencias.**

In [19]:
query = "SELECT * FROM Customer WHERE SupportRepId >= 3 and SupportRepId <= 4"

cursor.execute(query)
incidence = cursor.fetchall()

incidence_df = pd.DataFrame(incidence, columns=column_names)
incidence_df

CustomerId  FirstName     LastName  \
0            1       Luís    Gonçalves   
1            3   François     Tremblay   
2           12    Roberto      Almeida   
3           15   Jennifer     Peterson   
4           18   Michelle       Brooks   
5           19        Tim        Goyer   
6           24      Frank      Ralston   
7           29     Robert        Brown   
8           30     Edward      Francis   
9           33      Ellie     Sullivan   
10          37       Fynn   Zimmermann   
11          38     Niklas     Schröder   
12          42      Wyatt       Girard   
13          43   Isabelle      Mercier   
14          44      Terhi   Hämäläinen   
15          45   Ladislav       Kovács   
16          46       Hugh     O'Reilly   
17          52       Emma        Jones   
18          53       Phil       Hughes   
19          58      Manoj       Pareek   
20          59       Puja   Srivastava   
21           4      Bjørn       Hansen   
22           5  František  Wichterlová   
23           8       Daan      Peeters   
24           9       Kara      Nielsen   
25          10    Eduardo      Martins   
26          13   Fernanda        Ramos   
27          16      Frank       Harris   
28          20        Dan       Miller   
29          22    Heather      Leacock   
30          23       John       Gordon   
31          26    Richard   Cunningham   
32          27    Patrick         Gray   
33          32      Aaron     Mitchell   
34          34       João    Fernandes   
35          35   Madalena      Sampaio   
36          39    Camille      Bernard   
37          40  Dominique     Lefebvre   
38          49  Stanisław       Wójcik   
39          55       Mark       Taylor   
40          56      Diego    Gutiérrez   

                                             Company  \
0   Embraer - Empresa Brasileira de Aeronáutica S.A.   
1                                               None   
2                                             Riotur   
3                                      Rogers Canada   
4                                               None   
5                                         Apple Inc.   
6                                               None   
7                                               None   
8                                               None   
9                                               None   
10                                              None   
11                                              None   
12                                              None   
13                                              None   
14                                              None   
15                                              None   
16                                              None   
17                                              None   
18                                              None   
19                                              None   
20                                              None   
21                                              None   
22                                  JetBrains s.r.o.   
23                                              None   
24                                              None   
25                                  Woodstock Discos   
26                                              None   
27                                       Google Inc.   
28                                              None   
29                                              None   
30                                              None   
31                                              None   
32                                              None   
33                                              None   
34                                              None   
35                                              None   
36                                              None   
37                                              None   
38                                              None   

## Actividad 3-3 (Limpieza de personal)
Usaremos la tabla Employee (empleados).
1. Cuenta la cantidad de empleados que hay por ciudad.
2. Cuenta la cantidad de empleados que hay por departamento.

Subir Nota:
1. Muestra la edad de cada empleado.
2. Calcula la media de edad por departamento (Title).
3. Cuenta los empleados que fueron contratados, de media, por año

**1. Cuenta la cantidad de empleados que hay por ciudad.**

In [20]:
query = """
SELECT City, COUNT(*) as EmployeeCount 
FROM Employee 
GROUP BY City;
"""

cursor.execute(query)
employee = cursor.fetchall()

column_names= [description[0] for description in cursor.description]

employee_df = pd.DataFrame(employee, columns=column_names)
employee_df

City  EmployeeCount
0     Calgary              5
1    Edmonton              1
2  Lethbridge              2

**2. Cuenta la cantidad de empleados que hay por departamento.**

In [21]:
query = """
SELECT Title, COUNT(*) as Departament 
FROM Employee 
GROUP BY Title;
"""
cursor.execute(query)
department = cursor.fetchall()

column_names= [description[0] for description in cursor.description]

department_df = pd.DataFrame(department, columns=column_names)
department_df

Title  Departament
0      General Manager            1
1           IT Manager            1
2             IT Staff            2
3        Sales Manager            1
4  Sales Support Agent            3

Nota

**1. Muestra la edad de cada empleado**

In [22]:
query = """
SELECT EmployeeId, FirstName, LastName, Title, (strftime('%Y', 'now') - strftime('%Y', BirthDate)) as Age
FROM Employee;
"""

cursor.execute(query)
age = cursor.fetchall()

column_names= [description[0] for description in cursor.description]

age_df = pd.DataFrame(age, columns=column_names)
age_df

EmployeeId FirstName  LastName                Title  Age
0           1    Andrew     Adams      General Manager   62
1           2     Nancy   Edwards        Sales Manager   66
2           3      Jane   Peacock  Sales Support Agent   51
3           4  Margaret      Park  Sales Support Agent   77
4           5     Steve   Johnson  Sales Support Agent   59
5           6   Michael  Mitchell           IT Manager   51
6           7    Robert      King             IT Staff   54
7           8     Laura  Callahan             IT Staff   56

Nota 

**2. Calcula la media de edad por departamento (Title).**

In [23]:
query = """
SELECT Title, AVG((strftime('%Y', 'now') - strftime('%Y', BirthDate))) AS AvgAge
FROM Employee
GROUP BY Title;
"""

cursor.execute(query)
avg_age = cursor.fetchall()

column_names= [description[0] for description in cursor.description]

avg_age_df = pd.DataFrame(avg_age, columns=column_names)
avg_age_df

Title     AvgAge
0      General Manager  62.000000
1           IT Manager  51.000000
2             IT Staff  55.000000
3        Sales Manager  66.000000
4  Sales Support Agent  62.333333

Nota

**3. Cuenta los empleados que fueron contratados, de media, por año**

In [24]:
query = """
SELECT strftime('%Y', HireDate) AS HireYear, COUNT(*) AS NumEmployees
FROM Employee
GROUP BY HireYear;
"""
cursor.execute(query)
hired = cursor.fetchall()

column_names = [description[0] for description in cursor.description]
hired_df = pd.DataFrame(hired, columns=column_names)
hired_df

HireYear  NumEmployees
0     2002             3
1     2003             3
2     2004             2

## Actividad 4-1 (Experto en canciones)
Volvemos a usar la tabla Track (canción).
1. Muestra todas las canciones con el MediaType Protected AAC audio file.
2. Muestra todas las canciones que contengan algún MediaType con AAC.
3. Muestra todas las canciones que duren más de 2 minutos.
4. Muestra todas las canciones de Jazz.
5. Averigua cual es la canción más pesada.

Subir Nota:
1. ¿Cuantos discos tiene Led Zeppelin?
2. De entre sus discos, ¿cuanto cuesta el disco Houses Of The Holy?

**1. Muestra todas las canciones con el MediaType Protected AAC audio file.**

In [25]:
query = "SELECT * FROM MediaType"
cursor.execute(query)
media_type= cursor.fetchall()

column_names = [description[0] for description in cursor.description]
media_type = pd.DataFrame(media_type, columns=column_names)
media_type

MediaTypeId                         Name
0            1              MPEG audio file
1            2     Protected AAC audio file
2            3  Protected MPEG-4 video file
3            4     Purchased AAC audio file
4            5               AAC audio file

In [26]:
# Note:  MediaTypeId 2 → Protected AAC audio file
query = """
SELECT Track.Name AS Song, MediaType.Name AS MediaType
FROM Track
JOIN MediaType ON Track.MediaTypeId = MediaType.MediaTypeId
WHERE MediaType.Name = 'Protected AAC audio file'
"""
cursor.execute(query)
songs = cursor.fetchall()

column_names = [description[0] for description in cursor.description]
songs_df = pd.DataFrame(songs, columns=column_names)
songs_df

Song  \
0                                    Balls to the Wall   
1                                      Fast As a Shark   
2                                    Restless and Wild   
3                                 Princess of the Dawn   
4                                Welcome to the Jungle   
..                                                 ...   
232  Pini Di Roma (Pinien Von Rom) \ I Pini Della V...   
233  String Quartet No. 12 in C Minor, D. 703 "Quar...   
234               L'orfeo, Act 3, Sinfonia (Orchestra)   
235  Quintet for Horn, Violin, 2 Violas, and Cello ...   
236                                      Koyaanisqatsi   

                    MediaType  
0    Protected AAC audio file  
1    Protected AAC audio file  
2    Protected AAC audio file  
3    Protected AAC audio file  
4    Protected AAC audio file  
..                        ...  
232  Protected AAC audio file  
233  Protected AAC audio file  
234  Protected AAC audio file  
235  Protected AAC audio file  
236  Protected AAC audio file  

[237 rows x 2 columns]

**2. Muestra todas las canciones que contengan algún MediaType con AAC.**

In [27]:
query = """
SELECT Track.Name AS Song, MediaType.Name AS MediaType
FROM Track
JOIN MediaType ON Track.MediaTypeId = MediaType.MediaTypeId
WHERE MediaType.Name LIKE '%AAC%'
"""
cursor.execute(query)
songs_acc = cursor.fetchall()

column_names = [description[0] for description in cursor.description]
songs_acc_df = pd.DataFrame(songs_acc, columns=column_names)
songs_acc_df

Song  \
0                                    Balls to the Wall   
1                                      Fast As a Shark   
2                                    Restless and Wild   
3                                 Princess of the Dawn   
4                                Welcome to the Jungle   
..                                                 ...   
250  Pini Di Roma (Pinien Von Rom) \ I Pini Della V...   
251  String Quartet No. 12 in C Minor, D. 703 "Quar...   
252               L'orfeo, Act 3, Sinfonia (Orchestra)   
253  Quintet for Horn, Violin, 2 Violas, and Cello ...   
254                                      Koyaanisqatsi   

                    MediaType  
0    Protected AAC audio file  
1    Protected AAC audio file  
2    Protected AAC audio file  
3    Protected AAC audio file  
4    Protected AAC audio file  
..                        ...  
250  Protected AAC audio file  
251  Protected AAC audio file  
252  Protected AAC audio file  
253  Protected AAC audio file  
254  Protected AAC audio file  

[255 rows x 2 columns]

**3. Muestra todas las canciones que duren más de 2 minutos.**

In [28]:
query = """
SELECT Name AS Songs, Milliseconds / 60000.0 AS Minutes
FROM Track
WHERE Minutes >= 2
"""

cursor.execute(query)
two_min = cursor.fetchall()

column_names = [description[0] for description in cursor.description]
two_min_df = pd.DataFrame(two_min, columns=column_names)
two_min_df

Songs   Minutes
0               For Those About To Rock (We Salute You)  5.728650
1                                     Balls to the Wall  5.709367
2                                       Fast As a Shark  3.843650
3                                     Restless and Wild  4.200850
4                                  Princess of the Dawn  6.256967
...                                                 ...       ...
3405  Concerto for Violin, Strings and Continuo in G...  8.226217
3406  Pini Di Roma (Pinien Von Rom) \ I Pini Della V...  4.779017
3407  String Quartet No. 12 in C Minor, D. 703 "Quar...  2.320000
3408  Quintet for Horn, Violin, 2 Violas, and Cello ...  3.688850
3409                                      Koyaanisqatsi  3.433417

[3410 rows x 2 columns]

**4. Muestra todas las canciones de Jazz.**

In [29]:
query = "SELECT * FROM Genre"
cursor.execute(query)
genre_type= cursor.fetchall()

genre_type = pd.DataFrame(genre_type, columns=["GenreId", "Name"])
genre_type.head()

GenreId                Name
0        1                Rock
1        2                Jazz
2        3               Metal
3        4  Alternative & Punk
4        5       Rock And Roll

In [30]:
# Note:  GenreId 2 → Jazz
query = """
SELECT Track.Name AS Song, Genre.Name AS MusicalGenre
FROM Track
JOIN Genre ON Track.GenreId = Genre.GenreId
WHERE MusicalGenre = 'Jazz'
"""
cursor.execute(query)
jazz= cursor.fetchall()

column_names = [description[0] for description in cursor.description]
jazz = pd.DataFrame(jazz, columns=["GenreId", "Name"])
jazz

GenreId  Name
0                               Desafinado  Jazz
1                        Garota De Ipanema  Jazz
2    Samba De Uma Nota Só (One Note Samba)  Jazz
3                        Por Causa De Você  Jazz
4                                    Ligia  Jazz
..                                     ...   ...
125                           Little Linda  Jazz
126                     End Of Romanticism  Jazz
127                                 Amanda  Jazz
128                              Despertar  Jazz
129                            OAM's Blues  Jazz

[130 rows x 2 columns]

**5. Averigua cual es la canción más pesada**

In [31]:
query = "SELECT Name, MAX(Bytes) FROM Track"
cursor.execute(query)
song= cursor.fetchall()

song = pd.DataFrame(song, columns=["Song", "Heaviest song"])
song

Song  Heaviest song
0  Through a Looking Glass     1059546140

Subir Nota:

**1. ¿Cuantos discos tiene Led Zeppelin?**

In [32]:
query = """
SELECT COUNT(DISTINCT Album.AlbumId) AS "Led Zeppelin discs"
FROM Album
JOIN Artist ON Album.ArtistId = Artist.ArtistId
WHERE Artist.Name = 'Led Zeppelin'
"""
cursor.execute(query)
result= cursor.fetchall()

column_names = [description[0] for description in cursor.description]
result = pd.DataFrame(result, columns=column_names)
result

Led Zeppelin discs
0                  14

Subir Nota:

**2. De entre sus discos, ¿cuanto cuesta el disco Houses Of The Holy?**

In [33]:
query = """
SELECT Album.Title, SUM(Track.UnitPrice) AS "Total Cost"
FROM Album
JOIN Track ON Album.AlbumId = Track.AlbumId
WHERE Album.Title = 'Houses Of The Holy'
GROUP BY Album.Title
"""
cursor.execute(query)
result= cursor.fetchall()

column_names = [description[0] for description in cursor.description]
result = pd.DataFrame(result, columns=column_names)
result

Title  Total Cost
0  Houses Of The Holy        7.92

## Actividad 5-1 (Solo compañías)
De la tabla Customer crea una vista llamada Customer_with_companies, donde estarán incluidos todos los resultados salvo cuando Company sea NULL. A partir de la vista realiza las siguientes acciones.
1. Ordena los resultados por orden alfabético de Company.
2. Muestra que compañías son de Brazil.

In [35]:
query = """
CREATE TEMPORARY VIEW Customer_with_companies AS SELECT * FROM Customer
WHERE Company IS NOT NULL
"""
cursor.execute(query)

OperationalError: view Customer_with_companies already exists

**1. Ordena los resultados por orden alfabético de Company.**

In [36]:
# Fetch data from the temporary view
cursor.execute("SELECT * FROM Customer_with_companies ORDER BY Company ASC")
result = cursor.fetchall()

column_names = [description[0] for description in cursor.description]
result = pd.DataFrame(result, columns=column_names)
result

CustomerId  FirstName     LastName  \
0          19        Tim        Goyer   
1          11  Alexandre        Rocha   
2           1       Luís    Gonçalves   
3          16      Frank       Harris   
4           5  František  Wichterlová   
5          17       Jack        Smith   
6          12    Roberto      Almeida   
7          15   Jennifer     Peterson   
8          14       Mark      Philips   
9          10    Eduardo      Martins   

                                            Company  \
0                                        Apple Inc.   
1                              Banco do Brasil S.A.   
2  Embraer - Empresa Brasileira de Aeronáutica S.A.   
3                                       Google Inc.   
4                                  JetBrains s.r.o.   
5                             Microsoft Corporation   
6                                            Riotur   
7                                     Rogers Canada   
8                                             Telus   
9                                  Woodstock Discos   

                           Address                 City State         Country  \
0                  1 Infinite Loop            Cupertino    CA             USA   
1               Av. Paulista, 2022            São Paulo    SP          Brazil   
2  Av. Brigadeiro Faria Lima, 2170  São José dos Campos    SP          Brazil   
3        1600 Amphitheatre Parkway        Mountain View    CA             USA   
4                    Klanova 9/506               Prague  None  Czech Republic   
5                  1 Microsoft Way              Redmond    WA             USA   
6                 Praça Pio X, 119       Rio de Janeiro    RJ          Brazil   
7              700 W Pender Street            Vancouver    BC          Canada   
8                   8210 111 ST NW             Edmonton    AB          Canada   
9        Rua Dr. Falcão Filho, 155            São Paulo    SP          Brazil   

   PostalCode               Phone                 Fax  \
0       95014   +1 (408) 996-1010   +1 (408) 996-1011   
1   01310-200  +55 (11) 3055-3278  +55 (11) 3055-8131   
2   12227-000  +55 (12) 3923-5555  +55 (12) 3923-5566   
3  94043-1351   +1 (650) 253-0000   +1 (650) 253-0000   
4       14700    +420 2 4172 5555    +420 2 4172 5555   
5  98052-8300   +1 (425) 882-8080   +1 (425) 882-8081   
6   20040-020  +55 (21) 2271-7000  +55 (21) 2271-7070   
7     V6C 1G8   +1 (604) 688-2255   +1 (604) 688-8756   
8     T6G 2C7   +1 (780) 434-4554   +1 (780) 434-5565   
9   01007-010  +55 (11) 3033-5446  +55 (11) 3033-4564   

                           Email  SupportRepId  
0               tgoyer@apple.com             3  
1               alero@uol.com.br             5  
2           luisg@embraer.com.br             3  
3             fharris@google.com             4  
4       frantisekw@jetbrains.com             4  
5        jacksmith@microsoft.com             5  
6  roberto.almeida@riotur.gov.br             3  
7            jenniferp@rogers.ca             3  
8             mphilips12@shaw.ca             5  
9       eduardo@woodstock.com.br             4

**2. Muestra que compañías son de Brazil.**

In [37]:
cursor.execute("SELECT Company FROM Customer_with_companies WHERE Country == 'Brazil'")
result = cursor.fetchall()

column_names = [description[0] for description in cursor.description]
result = pd.DataFrame(result, columns=column_names)
result

Company
0  Embraer - Empresa Brasileira de Aeronáutica S.A.
1                                  Woodstock Discos
2                              Banco do Brasil S.A.
3                                            Riotur

## Actividad 6-1 (Dame la factura)
De la tabla Track, consigue la siguiente información.
1. Cual es el título de la canción que menos pesa (Bytes).
2. Cual es el título de la canción que más dura (Miliseconds).
3. Cuantas canciones cuestan 1$ o más.
4. Cuantas canciones hay de Queen.
5. Cual es la media de duración entre todas las canciones.
6. Cual es la media de peso entre todas las canciones de U2.
7. Cuantas canciones esta Bill Berry como Composer (Compositor).
8. Un Mb son: Bite / 1024 / 1024. Muestra todos los Tracks calculando, y renombrando, la columna Bytes en Mb.

In [39]:
# 1. Cual es el título de la canción que menos pesa (Bytes).
query = "SELECT NAME, MIN(Bytes) FROM Track"
cursor.execute(query)
result = cursor.fetchall()

column_names = [description[0] for description in cursor.description]
result = pd.DataFrame(result, columns=column_names)
result

Name  MIN(Bytes)
0  É Uma Partida De Futebol       38747

In [40]:
# 2. Cual es el título de la canción que más dura (Miliseconds).
query = "SELECT NAME, MAX(Milliseconds) FROM Track"
cursor.execute(query)
result = cursor.fetchall()

column_names = [description[0] for description in cursor.description]
result = pd.DataFrame(result, columns=column_names)
result

Name  MAX(Milliseconds)
0  Occupation / Precipice            5286953

In [41]:
# 3. Cuantas canciones cuestan 1$ o más.
query = "SELECT COUNT(Name) AS 'Songs > 1$' FROM Track WHERE UnitPrice > 1"
cursor.execute(query)
result = cursor.fetchall()

column_names = [description[0] for description in cursor.description]
result = pd.DataFrame(result, columns=column_names)
result

Songs > 1$
0         213

In [42]:
# 4. Cuantas canciones hay de Queen.
query = "SELECT COUNT(*) AS 'Queen Songs' FROM Track WHERE Composer LIKE '%Queen%'"
cursor.execute(query)
result = cursor.fetchall()

column_names = [description[0] for description in cursor.description]
result = pd.DataFrame(result, columns=column_names)
result

Queen Songs
0           10

In [43]:
# 5. Cual es la media de duración entre todas las canciones.
query = "SELECT AVG(Milliseconds) AS 'Average Time (milisec)' FROM Track"
cursor.execute(query)
result = cursor.fetchall()

column_names = [description[0] for description in cursor.description]
result = pd.DataFrame(result, columns=column_names)
result

Average Time (milisec)
0           393599.212104

In [44]:
# 6. Cual es la media de peso entre todas las canciones de U2.
query = "SELECT AVG(Bytes) AS 'Average weight (bytes)' FROM Track WHERE Composer LIKE '%U2%'"
cursor.execute(query)
result = cursor.fetchall()

column_names = [description[0] for description in cursor.description]
result = pd.DataFrame(result, columns=column_names)
result

Average weight (bytes)
0            8.844166e+06

In [ ]:
# 7. Cuantas canciones esta Bill Berry como Composer (Compositor).

In [ ]:
# 8. Un Mb son: Bite / 1024 / 1024. Muestra todos los Tracks calculando, y renombrando, la columna Bytes en Mb.

## Actividad 7-1 (Buscando facturas)
De la tabla Invoice, consigue la siguiente información.
1. Muestra: InvoiceId, nombre del cliente y BillingCountry.
2. Ordena de mayor a menor por Total.
3. Cual es el país que más a facturado.

ado.

## Actividad 8-1 (Alimentado la productora)
1. Añade 5 artistas que te gusten en la tabla Artist.
2. Introduce el MediaType Wav.
3. Crea 2 discos que estén relacionados con los artistas que has creado.

 Track.

## Actividad 9-1 (Artistas desaparecidos)
Volvemos a usar la tabla Artist (canción). (usa PRAGMA foreign_keys = OFF para desactivar la comprobación)
1. Borra a U2.
2. Borra los artistas que tengan en su nombre el símbolo &.
3. Borra los artistas con una Id entre 201 y 230.
4. Borra toda la tabla de Track.

ncerto.

## Actividad 10-1 (Se acabaron las rebajas)
Volvemos a usar la tabla Track (canción).
1. Sube el precio de todas las canciones de 0.99 a 2.99.
2. Sube el precio de todas las canciones de 1.99 a 4.99.
3. Cambia el Compositor (Composer) por Sara del TrackId número 2000.
4. Cambia el Compositor (Composer) por clasico de todas las canciones que empiecen con el nombre Concerto.